lightweight editorial suggestions and lose "canonical" (#428)
ietf-wg-jsonpath/draft-ietf-jsonpath-base: draft-ietf-jsonpath-base.md

: Formal parameter that can take Arguments (actual parameters) in a function-expression.

Normalized Path:

> : A minimal form of JSONPath expression that identifies a node in a value by

Reviewer: I'm not sure this is right either.  If anything, the "normalized" syntax is more verbose for many cases (e.g. `$.a.b` vs `$['a']['b']`.

Maybe "formal"?

Reviewer 2: Not sure "formal" is a great deal better. Perhaps just skip the adjective altogether?
```suggestion
: A form of JSONPath expression that identifies a node in a value by
```

Reviewer 3: > Not sure "formal" is a great deal better. Perhaps just skip the adjective altogether?

Uniquely?

Reviewer 2: I think uniqueness of the identified node is implicit. Or did you mean uniqueness of the query??

Reviewer 3: For each node, there is exactly one ("unique") normalized path.
That is the main point of them to me.

Reviewer 2: Get rid of minimal now we have expanded the definition:
```suggestion
: A form of JSONPath expression that identifies a node in a value by
```

>   providing a query that results in exactly that node.  Similar

Reviewer 3: So let's do the "unique" here:

```suggestion
  providing a unique query that results in exactly that node.  Similar
```

Reviewer: Here, "unique" doesn't work.  Identifying `foo` in `{"foo": "bar"}` can be achieved using `$['foo']` and `$.foo`.  The key here is that no shorthand syntaxes are used.

Secondly, are Normalized Paths necessarily Singular Paths?  If so, we should state that.

I think these two requirements (Singular Path and expressed without shorthand) are sufficient to define Normalized Path.

Reviewer 2: > Here, "unique" doesn't work. Identifying `foo` in `{"foo": "bar"}` can be achieved using `$['foo']` and `$.foo`. The key here is that no shorthand syntaxes are used.

The way I read the spec is that the terminology section gives a loose definition of Normalized Path which is tightened up in section 2.9. With that reading, it would be ok to add uniqueness to the term definition because that falls out of the tighter definition in section 2.9. Maybe we need to add a forward reference from the term definition to section 2.9 to make it clear this is the way to read the spec? (I wouldn't want to add the full definition of Normalized Path to the terminology section. )

> 
> Secondly, are Normalized Paths necessarily Singular Paths? If so, we should state that.

Yes, they are. Section 2.9 already says:

>> Note: Normalized Paths are Singular Queries, but not all Singular Queries are Normalized Paths.

> 
> I think these two requirements (Singular Path and expressed without shorthand) are sufficient to define Normalized Path.

Unfortunately, those conditions are not sufficient. As section 2.9 says:

>> For example, $[-3] is a Singular Query, but is not a Normalized Path.

`$[-3]` is a Singular Query and is expressed without shorthand. (NB. the spec doesn't describe such array indices as shorthand.)

(PS. Pardon the above markup. I used `>>` for quotes from the draft to distinguish them from quotes of the earlier comment.)

Reviewer: All of that is fair, I suppose.

The thing with `$[-3]` is that it's not unique _for all JSON instances_.  Processing that path on JSON arrays of different lengths will select nodes at different positions, but processing `$[3]` on those arrays will always select the fourth node (if it exists).

Maybe

> providing a unique query that results in exactly that node for all JSON instances.

(or "data" or whatever instead of "instances" [ref](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/427))

Reviewer 2: Ok, how about `$[1:2]` (as a synonym for `$[1]`)? It's a Singular Query, without shorthand, and results in the same node for all JSON instances.

Reviewer: Yeah, I don't know.  This is hard to nail down, and it's late for me. 🙂

Reviewer 3: There is nothing that is fundamentally better about always using an unsigned integer to represent normalized paths than using a negative integer.
We just decided always using the unsigned integer integer is more appealing (probably because we feel that this is a more fundamental way to access array elements, but that is not fundamentally so...).
So normalized paths are first and foremost a convention that lead to them being unique for each node.

Reviewer 3: We could be more explicit and add a sentence:
Each node in an argument is identified by exactly one normalize path (we say, the normalized path is "unique" for that node), and, to be a normalized path for a specific argument, the normalize path needs to identify exactly one node.

Reviewer 2: ```suggestion
  providing a query that results in exactly that node.  Each node in an argument is identified by exactly one Normalized Path (we say, the Normalized Path is "unique" for that node), and, to be a Normalized Path for a specific argument, the Normalized Path needs to identify exactly one node. Similar
```

i.e., to the argument as a whole.

The current node identifier `@` refers to the current node in the context

> of the evaluation of a filter expression (described below).

Reviewer: or just link to the section?

Reviewer 3: Yep.  One more, later.

> of the evaluation of a filter expression (described below).

Reviewer 3: ```suggestion
of the evaluation of a filter expression ({{filter-selector}}).
```

### Selectors

> A name selector, e.g. `'name'`, selects a named child in an object.

Reviewer 2: We can talk about a named value in an object, but when it comes to children, I would say that an object's children are children _of_ the object, not children _in_ the object. Similarly for arrays.

Reviewer 3: Unborn child?  (Just kidding.)
+1

Reviewer 2: Ok:
```suggestion
A name selector, e.g. `'name'`, selects a named child of an object.
```

Reviewer 3: Since "of" includes "in" (the keyboard *of* my laptop), I think we can accept this suggestion.

> An index selector, e.g. `3`, selects an indexed child in an array.

Reviewer 2: Similarly:
```suggestion
An index selector, e.g. `3`, selects an indexed child of an array.
```

~~~~

The slice selector consists of three optional decimal integers separated by colons.

> The second colon is omitted when the third integer is.

Reviewer 2: Not so. `::` is a valid slice for example.

> The second colon is omitted when the third integer is.

Reviewer 3: ```suggestion
The second colon can be omitted when the third integer is.
```

Comparison expressions are available for comparisons between primitive
values (that is, numbers, strings, `true`, `false`, and `null`).
These can be obtained via literal values; Singular Queries, each of
which selects at most one node the value of which is then used; or
function expressions (see {{fnex}}) of type `ValueType` or

> `NodesType` (see {{type-conv}}).

Reviewer: I think with the addition of `value()`, `NodesType` is no longer valid in a comparison.  Is this an oversight?

Reviewer 2: Good catch. I think this oversight is repeated in 2.6.3: "If it occurs directly as a comparable in a comparison, the function is declared to have a result type of ValueType, or (conversion applies) NodesType." doesn't need to talk about `NodesType`.

Reviewer: Addressed in #429.

The order in which the children of an object appear in the resultant nodelist is not stipulated,
since JSON objects are unordered.

> Children of an array are ordered by their array index in the resultant nodelist.

Reviewer: We may want to avoid "index" outside of the index selector.

> Children of an array are ordered by their array index in the resultant nodelist.

Reviewer 3: ```suggestion
In the resultant nodelist, children of an array are ordered by their position in the array.
```

## Normalized Paths

> A Normalized Path is a minimal representation of the location of a node in a value which

Reviewer: "formal" here too, maybe?

Reviewer 2: Again:
```suggestion
A Normalized Path is a representation of the location of a node in a value which
```

> A Normalized Path is a minimal representation of the location of a node in a value which

Reviewer 3: ```suggestion
A Normalized Path is a unique representation of the location of a node in a value which
```

This is definitely a place where "unique" fits.  (Not so great in the other place further up.)

uniquely identifies the node in the value.
Specifically, a Normalized Path is a JSONPath query with restricted syntax (defined below),
e.g., `$['book'][3]`, which when applied to the value results in a nodelist consisting
of just the node identified by the Normalized Path.
Note that a Normalized Path represents the identity of a node _in a specific value_.
There is precisely one Normalized Path identifying any particular node in a value.

> A nodelist may be represented compactly in JSON as an array of strings, where the strings are

Reviewer: Actually, looking at this again, I think the adjective is fine.  It's describing a nodelist, which is more compact if you just have the Normalized Paths without the values.

Rescinding my comment here.

Normalized Paths use the canonical bracket notation, rather than dot notation.

Single quotes are used in Normalized Paths to delimit string member names. This reduces the
number of characters that need escaping when Normalized Paths appear in double quote-delimited
strings, e.g., in JSON texts.

> Certain characters are escaped in Noramlized Paths, in one and only one way; all other

Reviewer 2: ```suggestion
Certain characters are escaped in Normalized Paths, in one and only one way; all other
```
